stackslib/src/chainstate/stacks/boot/signers-voting.clar

```diff
@@ -62,35 +81,96 @@
             )
         (get prepare-cycle-length pox-info)))
 
+;; get the aggregate public key for the given reward cycle (or none)
+(define-read-only (get-approved-aggregate-key (reward-cycle uint))
+    (map-get? aggregate-public-keys reward-cycle))
+
 (define-private (is-in-voting-window (height uint) (reward-cycle uint))
     (let ((last-cycle (unwrap-panic (contract-call? .signers get-last-set-cycle))))
         (and (is-eq last-cycle reward-cycle)
             (is-in-prepare-phase height))))
 
-(define-public (vote-for-aggregate-public-key (signer-index uint) (key (buff 33)) (round uint))
-    (let ((reward-cycle (+ u1 (burn-height-to-reward-cycle burn-block-height)))
-            (tally-key {reward-cycle: reward-cycle, round: round, aggregate-public-key: key})
-            ;; one slot, one vote
-            (num-slots (try! (get-current-signer-slots signer-index)))
-            (new-total (+ num-slots (default-to u0 (map-get? tally tally-key)))))
-        (asserts! (is-in-voting-window burn-block-height reward-cycle) err-out-of-voting-window)
-        (asserts! (>= round (default-to u0 (map-get? rounds reward-cycle))) err-old-round)
-        (asserts! (is-eq (len key) u33) err-ill-formed-aggregate-public-key)
-        (asserts! (is-valid-aggregated-public-key key {reward-cycle: reward-cycle, round: round}) err-duplicate-aggregate-public-key)
-        (asserts! (map-insert votes {reward-cycle: reward-cycle, round: round, signer: tx-sender} {aggregate-public-key: key, reward-slots: num-slots}) err-duplicate-vote)
-        (map-set tally tally-key new-total)
-        (map-set used-aggregate-public-keys key {reward-cycle: reward-cycle, round: round})
-        (update-last-round reward-cycle round)
-        (print { 
-            event: "voted", 
-            signer: tx-sender, 
-            reward-cycle: reward-cycle, 
-            round: round, 
-            key: key, 
-            new-total: new-total })
-        (ok true)))
+(define-private (sum-weights (signer { signer: principal, weight: uint }) (acc uint))
+    (+ acc (get weight signer)))
+
+(define-private (get-total-weight (reward-cycle uint))
+    (match (map-get? cycle-total-weight reward-cycle)
+        total (ok total)
+        (let ((signers (unwrap! (contract-call? .signers get-signers reward-cycle) (err ERR_FAILED_TO_RETRIEVE_SIGNERS)))
+                (total (fold sum-weights signers u0)))
+            (map-set cycle-total-weight reward-cycle total)
+            (ok total))))
 
 (define-private (update-last-round (reward-cycle uint) (round uint))
-    (match (map-get? rounds reward-cycle)
-        last-round (and (> round last-round) (map-set rounds reward-cycle round))
-        (map-set rounds reward-cycle round)))
+    (ok (match (map-get? rounds reward-cycle)
+        last-round (begin
+            (asserts! (<= round (+ last-round u1)) (err ERR_INVALID_ROUND))
+            (and (> round last-round) (map-set rounds reward-cycle round)))
```

Comment:
Can this be an `(if ...)` expression instead?  I think it's more descriptive of what's actually happening.

Reply:
Refactored in 252a98cc9